common/types.go

```diff
@@ -168,3 +169,71 @@ type LogConfig struct {
 	FileCount   uint64
 	TimeTracker bool
 }
+
+type Block struct {
+	Id         string
+	StartIndex int64
+	EndIndex   int64
+	Size       int64
+}
+
+// list that holds blocks containing ids and corresponding offsets
+type BlockOffsetList struct {
+	BlockOffsetList []*Block //blockId to offset mapping
+	SmallFile       bool     // does it consist of blocks
+	Cached          bool     // is it cached?
+}
+
+func (bol BlockOffsetList) binarySearch(offset int64) (bool, int) {
+	lowerBound := 0
+	higherBound := len(bol.BlockOffsetList) - 1
+	for lowerBound <= higherBound {
+		middleIndex := (lowerBound + higherBound) / 2
+		// we found the starting block that changes are being applied to
+		if bol.BlockOffsetList[middleIndex].EndIndex > offset && bol.BlockOffsetList[middleIndex].StartIndex <= offset {
+			return true, middleIndex
+			// if the end index is smaller or equal then we need to increase our lower bound
+		} else if bol.BlockOffsetList[middleIndex].EndIndex <= offset {
+			lowerBound = middleIndex + 1
+			// if the start index is larger than the offset we need to decrease our upper bound
+		} else if bol.BlockOffsetList[middleIndex].StartIndex > offset {
+			higherBound = middleIndex - 1
+		}
+	}
+	return false, 0
+}
+
+func (bol BlockOffsetList) FindBlocksToModify(offset, length int64) (*BlockOffsetList, int64, bool) {
+	// size of mod block list
+	size := int64(0)
+	currentBlockOffset := offset
+	modBlockList := BlockOffsetList{}
+	found, index := bol.binarySearch(offset)
```

Comment:
Block Size for a given file will be fixed as in all blocks of a given file will be of same size. If that is true then there is no need to search the blocks. We can directly convert offset to index and see how many block we are going to create/overwrite.

Reply:
You can't guarantee that. Blocks in a file can be of variable lengths. If they used other SDKs etc to append to the file.

Reply:
is this possible that blocks of the same file are variable length in nature ? My understanding was that all blocks have to be of same size.

Reply:
Yes blocks can be of variable lengths